pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go

```diff
@@ -489,71 +491,121 @@ func (lbaas *LbaasV2) UpdateLoadBalancer(clusterName string, service *api.Servic
-					}
-				}
-
-				return true, nil
-			})
-			if err != nil {
-				return false, err
-			}
-		}
 		return true, nil
 	})
 	if err != nil {
 		return err
 	}
 
-	// Anything left in addrs is a new member that needs to be added to a pool
-	for addr := range addrs {
-		_, err := v2_pools.CreateAssociateMember(lbaas.network, poolID, v2_pools.MemberCreateOpts{
-			Address:      addr,
-			ProtocolPort: int(ports[0].NodePort),
-			SubnetID:     lbaas.opts.SubnetId,
-		}).Extract()
+	// Compose Set of member (addresses) that _should_ exist
+	addrs := map[string]bool{}
+	for _, host := range hosts {
+		addr, err := getAddressByName(lbaas.compute, host)
 		if err != nil {
 			return err
 		}
-		waitLoadbalancerActiveProvisioningStatus(lbaas.network, loadbalancer.ID)
+		addrs[addr] = true
 	}
 
+	// Remove members that are no longer required
+	for _, pool := range poolsLists {
+		err := v2_pools.ListAssociateMembers(lbaas.network, pool.ID, v2_pools.MemberListOpts{}).EachPage(func(page pagination.Page) (bool, error) {
+			membersList, err := v2_pools.ExtractMembers(page)
+			if err != nil {
+				return false, err
+			}
+			for _, member := range membersList {
+				if _, found := addrs[member.Address]; found {
+					// Member already exists, remove from update list
+					addrs[member.Address] = false
+				} else {
+					// Member needs to be deleted
+					waitLoadbalancerActiveProvisioningStatus(lbaas.network, loadbalancer.ID)
+					err = v2_pools.DeleteMember(lbaas.network, pool.ID, member.ID).ExtractErr()
+					if err != nil {
+						return false, err
+					}
+					waitLoadbalancerActiveProvisioningStatus(lbaas.network, loadbalancer.ID)
+				}
+			}
+			return true, nil
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	// Add new members for each port in corresponding listener/pool
+	waitLoadbalancerActiveProvisioningStatus(lbaas.network, loadbalancer.ID)
+	for _, port := range ports {
+		// Find appropriate listener for this port
+		var listenerID string
+		for _, listener := range listenersLists {
+			if listener.ProtocolPort == int(port.Port) && listener.Protocol == string(port.Protocol) {
+				listenerID = listener.ID
+				break
+			}
+		}
+		if listenerID == "" {
+			return fmt.Errorf("Loadbalancer %s does not contain required listener for port %d and protocol %s", loadBalancerName, port.Port, port.Protocol)
+		}
+
+		// Find appropriate pool for found listener
+		var poolID string
+		for _, pool := range poolsLists {
+			for _, currentListener := range pool.Listeners {
+				if listenerID == currentListener.ID {
+					poolID = pool.ID
+					break
+				}
+			}
+			if poolID != "" {
+				break
+			}
+		}
+		if poolID == "" {
+			return fmt.Errorf("Loadbalancer %s does not contain required pool for listener %s", loadBalancerName, listenerID)
+		}
+
+		// Create a member for each new host
+		for addr := range addrs {
+			if !addrs[addr] {
+				// skip existing member
+				continue
```

Comment:
Previously we only had one pool, so we were naturally only checking/updating the same pool.

You can see in [servicecontroller.go](https://github.com/kubernetes/kubernetes/blob/5645ca749b8dde4779422371c04eab6d18d93263/pkg/controller/service/servicecontroller.go#L621) that `UpdateLoadBalancer` is basically called repeatedly and is expected to eventually update the list of hosts correctly - so a transient error will cause the current update to return early, but then the function will be called again (and again..) and expected to still do the right thing with this half-failed state.

I'll send a PR against your branch..